hal/src/stm32f2xx/i2c_hal.c

```diff
@@ -181,54 +188,99 @@ static void HAL_I2C_SoftwareReset(HAL_I2C_Interface i2c)
     i2cMap[i2c]->prevEnding = I2C_ENDING_UNKNOWN;
 }
 
-void HAL_I2C_Init(HAL_I2C_Interface i2c, void* reserved)
+static bool HAL_I2C_Config_Is_Valid(const HAL_I2C_Config* config) {
+    if ((config == NULL) ||
+            (config->rx_buffer == NULL ||
+             config->rx_buffer_size == 0 ||
+             config->tx_buffer == NULL ||
+             config->tx_buffer_size == 0))
+    {
+        return false;
+    }
+
+    return true;
+}
+
+static bool HAL_I2C_Buffers_Initialized(HAL_I2C_Interface i2c) {
+    if (i2cMap[i2c]->rxBuffer == NULL ||
+            i2cMap[i2c]->rxBufferSize == 0 ||
+            i2cMap[i2c]->txBuffer == NULL ||
+            i2cMap[i2c]->txBufferSize == 0) {
+        return false;
+    }
+
+    return true;
+}
+
+int HAL_I2C_Init(HAL_I2C_Interface i2c, const HAL_I2C_Config* config)
 {
     if(i2c == HAL_I2C_INTERFACE1)
     {
         i2cMap[i2c] = &I2C_MAP[I2C1_D0_D1];
     }
-#if PLATFORM_ID == 10 // Electron
+#if PLATFORM_ID == PLATFORM_ELECTRON_PRODUCTION // Electron
     if(i2c == HAL_I2C_INTERFACE2 || i2c == HAL_I2C_INTERFACE1)
     {
-    	   // these are co-dependent so initialize both
-       i2cMap[HAL_I2C_INTERFACE1] = &I2C_MAP[I2C1_D0_D1];
-       i2cMap[HAL_I2C_INTERFACE2] = &I2C_MAP[I2C3_C4_C5];
+        // these are co-dependent so initialize both
+        i2cMap[HAL_I2C_INTERFACE1] = &I2C_MAP[I2C1_D0_D1];
+        i2cMap[HAL_I2C_INTERFACE2] = &I2C_MAP[I2C3_C4_C5];
     }
     else if(i2c == HAL_I2C_INTERFACE3)
     {
         i2cMap[i2c] = &I2C_MAP[I2C3_PM_SDA_SCL];
     }
-#endif
 
-#if PLATFORM_ID == 10
     // For now we only enable this for PMIC I2C bus
     if (i2c == HAL_I2C_INTERFACE3) {
         os_thread_scheduling(false, NULL);
         if (i2cMap[i2c]->mutex == NULL) {
             os_mutex_recursive_create(&i2cMap[i2c]->mutex);
+        } else {
+            // Already initialized
+            os_thread_scheduling(true, NULL);
+            return SYSTEM_ERROR_NONE;
         }
         HAL_I2C_Acquire(i2c, NULL);
         os_thread_scheduling(true, NULL);
     }
-#endif // PLATFORM_ID == 10
+#endif // PLATFORM_ID == PLATFORM_ELECTRON_PRODUCTION
 
+    // Initialize internal data structure
+    if (HAL_I2C_Config_Is_Valid(config))
+    {
+        i2cMap[i2c]->rxBuffer = config->rx_buffer;
+        i2cMap[i2c]->rxBufferSize = config->rx_buffer_size;
+        i2cMap[i2c]->txBuffer = config->tx_buffer;
+        i2cMap[i2c]->txBufferSize = config->tx_buffer_size;
+    } else if (!HAL_I2C_Buffers_Initialized(i2c)) {
+        i2cMap[i2c]->rxBuffer = (uint8_t*)malloc(I2C_BUFFER_LENGTH);
```

Comment:
This might return `nullptr`. Same to `i2cMap[i2c]->txBuffer`.

Reply:
Added assertion.